.net9 dark theme

gerhardol opened the pull request:
## Proposed changes

Theming with https://github.com/dotnet/winforms/pull/11857
This is not as good looking as what @NikolayXHD did for GE 3.x, but it follows the standard Windows dark mode (there is a dark+ theme variant that stretches the design and is slightly darker).
There are many limitations and issues in https://github.com/dotnet/winforms/pull/11857 still, most are supposedly planned to be fixed for .net10.
The standard theme is not modified by this PR.

This PR has a number of tweaks to the setup to make this mostly usable, there are a number of issues still, maybe some more can be handled in .net9 too.

A few noted issues - many the same in ge 3 dark mode

* Add some kind of disclaimer for dark mode - at least that it is experimental in .net9. Win10 (and Win11 23H2?) is not expected to fully work too. Probably just in release notes.
* tooltips and various other like disabled text boxes are bright, no dramatic issue [(also reported)](https://github.com/dotnet/winforms/issues/11954)
* Forms extending from GitExtensionsForm are much darker than the "modern" forms based on GitExtensionsDialog. This applies to for instance Settings and FindInFile, total of 20 forms. There is a difference for the light theme too, but the difference is bigger for dark mode. Certainly not in this PR.
* Some adapted colors are a little too muted, add specific colors? Cleanup the adaption of colors in general?
* Some application adopted colors do not look great, like syntax highlighting in the editor. The colors are adapted from the hardcoded highlighting per file type, there is no way to set specific colors (or use system colors?).

For .NET10?

* ListView group headers like for remotes and in the dashboard is unreadable. FileStatusList is changed in #12116. Tracked in https://github.com/dotnet/winforms/issues/11953  This will not be changed in this PR, await .NET10
The ListView unreadable text is very annoying but I believe the text is in such forms that this can be accepted.
There will need to more color tweaking.

* Contrast is generally quite low though, most of this in the WinForms colors. The dark+ theme has slightly better contrast, more like ge3 dark.
Some overrides were already implemented, like for https://github.com/dotnet/winforms/issues/11953

Mostly run on Win11, a few visuals on Win10

The darksilver theme was removed, it would be the same as the dark theme as system colors cannot be changed.
highcontrast_dark was removed, not needed.
Invariant theme system colors are hardcoded, required to adjust colors.

I will appreciate if someone that is better in design and override .NET painting (.NET and GUI is not my strength, this is a hobby!) joins.

Some of the commits in this PR may be separated to individual PRs.

## Screenshots <!-- Remove this section if PR does not change UI -->

Mostly updated with the latest changes

Should this show GE 3.x/5.1/DarkMode?

For now some visual style examples and issues

General look
![image](https://github.com/user-attachments/assets/bbe57ced-162f-40f6-a953-b8a0aa6cc576)

dark+
![image](https://github.com/user-attachments/assets/5bd00be8-344b-4cfa-97f3-651ed03cc175)

Ansi colors in dark, mostly readable
![image](https://github.com/user-attachments/assets/7288c707-ade4-43c4-bb01-1376dda02415)

dark+
![image](https://github.com/user-attachments/assets/39987d90-8c63-42d9-9b5d-fdcb9b3c1541)

Example of Flat style (as in the PR) vs Standard.
![image](https://github.com/user-attachments/assets/5e965c06-56bb-458b-a546-369c1c028a57) 
![image](https://github.com/user-attachments/assets/9394deee-f47a-4f1b-ae79-03741f1dbaf5)

White background for unreadable textboxes, ox for single line, not for commit textbox
![image](https://github.com/user-attachments/assets/cbd966c4-7308-4df3-b739-526e959ab415)

![image](https://github.com/user-attachments/assets/5a6d7ba9-e689-45ad-837b-da4f1d6e3677)

A few usages of ListView that is not FileStatusList use ListView header
Keep as this until .net10
![image](https://github.com/user-attachments/assets/23f7bbb5-f785-4fb9-9c7f-f906afe59360)

Disabled button text is hard to read
![image](https://github.com/user-attachments/assets/7923bb44-43e5-44e3-90a6-cf96a6048325)

Panels in dark+ are darker, but tab foreground is still SystemColors.Windows (and inactive tabs is not so far from the background). Not so off though.
![image](https://github.com/user-attachments/assets/e28cbdfd-4e31-41a4-bdc5-749505ee5eaa)

## Test methodology <!-- How did you ensure quality? -->

Used as primary interface

## Test environment(s) <!-- Remove any that don't apply -->

- Git Extensions 33.33.33
- Git 2.47.1.windows.1
- Microsoft Windows NT 10.0.26100.0
- .NET 9.0.0
- DPI 96dpi (no scaling)
- Portable: False
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 9.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```


## Merge strategy

Rebase merge (PR submitter must change the commit message for the last commit).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented:
> Should this be GE 3.x/5.1/DarkMode?

I think this should be a dedicated branch, until it's deemed satisfactory/release-ready. Dark mode could be enabled via the settings, couldn't it?

RussKie commented on src/app/GitExtUtils/GitUI/Theming/ThemeSettings.cs:
```diff
@@ -26,5 +26,8 @@ public ThemeSettings(
 
         public bool UseSystemVisualStyle { get; }
+
+        public bool IsDarkTheme =>
+            Theme.GetNonEmptyColor(KnownColor.Window).GetBrightness() < 0.5;
```
There's a dedicated API in Windows Forms for that, isn't there?

gerhardol replied:
This is to know if the theme is dark, this is what was there.
I considered adding a specific color just to enable or a special setting to enable. We want to enable dark mode separate from the system for some time, there will be many visuals.

If we go with .net9 this can be merged if the most severe issues are fixed as experimental, there may be designers helping...
But also that will take a long time to get to.

gerhardol commented:
> > Should this be GE 3.x/5.1/DarkMode?
> 
> I think this should be a dedicated branch, until it's deemed satisfactory/release-ready. Dark mode could be enabled via the settings, couldn't it?

The comment was for screenshots.
As in the other comment, the invariant theme is not changed by this pr.

RussKie commented on src/app/GitExtensions/Project.Publish.targets:
```diff
@@ -17,21 +17,6 @@
-    ============================================================
-  <Target Name="_PublishExtraDependecies">
-    <Copy SourceFiles="$(SolutionDir)\Externals\EasyHook\EasyHookDll\Build\Release\x64\EasyHook64.dll" DestinationFolder="$(PublishDir)" />
-    <Copy SourceFiles="$(SolutionDir)\Externals\EasyHook\EasyHookDll\Build\Release\x86\EasyHook32.dll" DestinationFolder="$(PublishDir)" />
```
Update wix as well?

gerhardol replied:
fixing

RussKie commented on src/app/GitExtUtils/GitUI/Theming/ColorHelper.cs:
```diff
@@ -138,60 +142,16 @@ double DistanceTo(KnownColor c2)
+                    .TransformHsl(l: l => l - amount);
 
-        public static bool IsLightTheme()
+        public static bool IsDarkTheme()
```
Why not use [`Application.IsDarkModeEnabled`](https://learn.microsoft.com/dotnet/api/system.windows.forms.application.isdarkmodeenabled)?

gerhardol replied:
The short answer is that ThemeModule.IsDarkTheme cannot be used in Dashboard (where this is used).
I have mostly used Application.IsDarkModeEnabled for consequences of using the experimental dark mode, like overriding buttons to FlatStyle and ThemeModule.IsDarkTheme for other color relatded changes.
It is not exactly the same.

If the concept of theming is narrowed to light/dark for system colors (as WinForms supports) and custom for GE colors, then only Application.IsDarkModeEnabled could be used in all situations.
Then I would like to suppress WFO5001 globally too. I really hope that dark mode is not removed even if all quirks are not fixed for .net10 ((just two issues that I know of remains so GE could live with the current support.) Thanks to .NET engineers that fixed this in the final hours...

gerhardol replied:
Change to use the Application API and suppress WFO5001 ?

gerhardol replied:
@RussKie Do you have any input in if dark mode is kept in .net10 and if so if it will be experimental?
Can we assume it is going forward and suppress WFO5001, implementing as it will be kept?

RussKie replied:
I honestly don't know, I've moved from the team few years ago. 

@merriemcgaw @klausloeffelmann are there any plans you're aware of pulling the plug on the dark theme?

gerhardol replied:
There are a number of issues in GH already , like https://github.com/dotnet/winforms/issues/11953
Some in this PR may not be reported, will do that.
I can open an umbrella issue too - nice to have for this PR to link to as well.
So far no issues reported, I wanted to have this finished first and that takes some time (I am not a GUI designer).

Is there a summary of what have been fixed so far?

Some issues have workarounds, we use ThemeFix to correct them.
These are likely still the most important to fix in WinForms, many see them.
The other that (so far) have affected GE are in this issue. It is definitely the ListView label that is the most important, but can rank them in the umbrella issue.

@KlausLoeffelmann Thanks for the efforts with this! There are many users that will appreciate the weekend you spent before .net9 code freeze without knowing the history.

RussKie replied:
Thank you, @KlausLoeffelmann. This is great to hear.

> Please also [note this feature](https://learn.microsoft.com/dotnet/api/system.drawing.systemcolors.usealternativecolorset?view=net-9.0#system-drawing-systemcolors-usealternativecolorset) for your efforts, which can be easily overlooked.

I'm finding the doc a little vague on the usefulness of this feature. Are there any samples showing the purpose of this?


> The other that (so far) have affected GE are in this issue. It is definitely the ListView label that is the most important, but can rank them in the umbrella issue.

This is probably the most visible (or better say "invisible"). Other than that, the readonly colour is the most visible gap:
![image](https://github.com/user-attachments/assets/4cd6679d-283a-48cd-83e9-aeca45f9693f)

KlausLoeffelmann replied:
You mean the `SystemColors` switch, right?

The idea behind is that it switches the whole color set, albeit not configurable, for the `SystemColors`. We can override in the WinForms runtime quite a few things to make edge case scenarios work. But there are other cases where controls are really unchangable dependent on `SystemColors`. And also, if you wrote your own implementations, and used `SystemColors`, you want to automatically apply that set of dark-mode colors with "invers contrast" so to say, so the DarkMode-migration effort is as minimal as possible.

It not the most spectacular thing alright, and surely not helpful in all the cases. But I just did want you guys not to miss it, since it is kind of hard to notice, and maybe it _can_ come in handy here and there.

And while we're at it:
We're forcing a control into the DarkMode scene with a certain theming option. We know that that's not a perfect solution, but we decided, before we do nothing, we do what's possible, since at this point it is more than a "fashion" thing. I know a whole series of Devs, like me, struggling with ADHD, and for me, and with my current Monitor setting, Dark Mode is a REAL A11Y improvement, but at the same time: Light Mode makes my setup no longer usable. I am getting headaches 10 minutes in. And I hear those things more and more in the community, so, to echo the approach: Let me try to make happen what's somehow possible, but we may still not end up with everything being perfect - just to set expectations right.

Also: We cannot service .NET 9, since we do not service Experimental Features. That is unfortunately true also for features, which worked and regressed - like the `ToolStripStatusBar`, which doesn't anymore apply the correct background color after an A11Y fix.

But ping me directly for workarounds, I might be able to help with, if you run into something, and should I not react in time, because I am way more than swamped currently, I am absolutely sure Igor has the MS-internal means to ping me in a way that I cannot not notice...

RussKie replied:
I suppose we can try to spin a .NET 10 build too...

gerhardol replied:
@russkie
Can I merge this before testing further?
It is a one line change to disable dark mode if needed...

RussKie replied:
Sure thing

RussKie commented on src/app/GitUI/CommandsDialogs/BrowseDialog/DashboardControl/Dashboard.Designer.cs:
```diff
@@ -46,7 +48,7 @@ private void InitializeComponent()
             // 
             pnlLeft.Anchor = AnchorStyles.Top | AnchorStyles.Bottom | AnchorStyles.Left | AnchorStyles.Right;
-            pnlLeft.BackColor = Color.FromArgb(((int)(((byte)(223)))), ((int)(((byte)(228)))), ((int)(((byte)(235)))));
+            pnlLeft.BackColor = Color.FromArgb(((int)(((byte)(223)))), ((int)(((byte)(228)))), ((int)(((byte)(235))))).AdaptBackColor();
```
We shouldn't be manually changing the designer files, as these will get overwritten once we open the form in the designer.

gerhardol replied:
In this case, the color assign in itself is removed too, so not a difference.
Normally, you need to revert every change by the designer anyway.
I am not sure if something like `Remove.ImageTransparentColor = Color.Magenta.AdaptBackColor();`even need to be changed.
It was easiest to change all occurrences for now, to find the usages.

I would have preferred to split this PR in several sequiential PRs, but there will be feedback back for some time.
There are some blockers (.net in AppVeyor, FileStatusList use TreeView and likely a 5.2 release) before this is ready to merge anyway. But I find this very usable in normal work.

RussKie commented on src/app/GitExtensions/Program.cs:
```diff
@@ -67,23 +67,6 @@ private static void Main()
-            };
-#endif
 
             HighDpiMouseCursors.Enable();
```
I was kind of expecting we'd be calling [`Application.SetColorMode API`](https://learn.microsoft.com/dotnet/api/system.windows.forms.application.setcolormode) on line 45 depending on the user selection:

```cs
Application.SetColorMode(SystemColorMode);
```

gerhardol replied:
I just added a TODO for the Dashboard, to be cleaned up. That works well enough for now.

RussKie reviewed (commented):
🚀

gerhardol commented:
> Dark mode could be enabled via the settings, couldn't it?

default could set darkmode if enabled for apps when this is no longer experimental. Not in this PR...

gerhardol commented:
Except for the dependent PRs and the discussion items mentioned in the description, I consider this to be ready for review.

RussKie commented:
This is likely the Windows Forms issue but adding here for completeness:
![image](https://github.com/user-attachments/assets/b47cd847-d939-4a8c-9b7d-db7d22e5b4e8)

gerhardol commented:
> The active selection text needs to be light: 
> E.g., when the app is out of focus, the text becomes legible: 

The selection color (active/inactive) as well as the authored highlight is a little too close, but brighter colors do not look well (the contrast is lower).
I prefer it this way, but maybe the dark vs dark+ themes can be separate?
Bright colors in some forms like settings/reset as in #12218 is OK for me as it is transistent use, not something you stare at "for hours". This will not be perfect.

> The branch name colours need to be lightened up as well (or perhaps the label background needs to be lightened up) to increase contrast ratio:

branch color and labels are themable, this can be adjusted too, also separate dark/dark+?
The current values are a little too close, but this was what I found to be best looking.
(The yellow authored highlight was uglier but easier to handle.)

> This is likely the Windows Forms issue but adding here for completeness.
> 
> The listview group's text and the action's text aren't legible:

Yes, no workaround found, listed in the PR description.
The reason why mstv's ListView -> ListTree was required first (as most important use of these labels was in FileStatusList).

RussKie commented:
Trying to create a fixup:
![image](https://github.com/user-attachments/assets/ba2afcbd-ff27-4815-8b87-cf41cdcf5d13)

gerhardol commented:
> Trying to create a fixup:

Also in the description, read-only is white and no simple way to override

gerhardol commented:
Rebased and arranged the minor preparing commits first
To update description with read-only

> The active selection text needs to be light:

Not changed, that is SystemColors.Highlight.
Maybe implemet explorer style selection style (both grid and file status).
Not n this PR

> E.g., when the app is out of focus, the text becomes legible:

Changed to be lighter, less like in explorer style but diff to authored highlight is greater

> The branch name colours need to be lightened up as well (or perhaps the label background needs to be lightened up) to increase contrast ratio: 

Changed

![image](https://github.com/user-attachments/assets/8f365fa5-1302-4dac-b2db-0b084d061575)

RussKie commented:
I'm making call outs for the issues I stumble upon.
I'll dig in when I get some spare time.

On Thu, 27 Feb 2025 at 01:38, Gerhard Olsson ***@***.***>
wrote:

> The remotes (red) needs further tweaking.
>
> This is set in dark.css.
> Instead of me just guessing what you want. Can you set a value?
> Or just suggest a color code? with any service
>
> .RemoteBranch { color: #ff8080; } /* hsl(0, 100%, 75%) */
> changed to
> .RemoteBranch { color: #FF2626; }
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/12111#issuecomment-2685233586>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXS2ORZWDBFY4C7GSGT2RXGVTAVCNFSM6AAAAABUCDHWHGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBVGIZTGNJYGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
> [image: gerhardol]*gerhardol* left a comment
> (gitextensions/gitextensions#12111)
> <https://github.com/gitextensions/gitextensions/pull/12111#issuecomment-2685233586>
>
> The remotes (red) needs further tweaking.
>
> This is set in dark.css.
> Instead of me just guessing what you want. Can you set a value?
> Or just suggest a color code? with any service
>
> .RemoteBranch { color: #ff8080; } /* hsl(0, 100%, 75%) */
> changed to
> .RemoteBranch { color: #FF2626; }
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/12111#issuecomment-2685233586>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBTEXS2ORZWDBFY4C7GSGT2RXGVTAVCNFSM6AAAAABUCDHWHGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBVGIZTGNJYGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

RussKie commented:
@gerhardol I played tinkered with the dak theming, and I propose the following changes:

```diff
diff --git a/src/app/GitUI/Themes/dark+.css b/src/app/GitUI/Themes/dark+.css
index b9821f754..79abc438d 100644
--- a/src/app/GitUI/Themes/dark+.css
+++ b/src/app/GitUI/Themes/dark+.css
@@ -6,6 +6,7 @@
 .PanelBackground { color: #252526; }
 .EditorBackground { color: #1e1e1e; }
 .LineNumberBackground { color: #303030; }
+.AuthoredHighlight { color: #373737; }
 
 .Branch { color: #00d76b; }
-.RemoteBranch { color: #ff8080; } /* hsl(0,   100%, 75%) */
+.RemoteBranch { color: #fd9797; } /* hsl(0, 100%, 70%) */
\ No newline at end of file
```

```diff
diff --git a/src/app/GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs b/src/app/GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs
index 5c8ae0f4d..ca9946dbf 100644
--- a/src/app/GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs
+++ b/src/app/GitUI/UserControls/RevisionGrid/RevisionDataGridView.cs
@@ -238,30 +238,36 @@ internal void AddColumn(ColumnProvider columnProvider)
         }
 
         private Color GetForeground(bool isSelected, bool isFocused, bool isNonRelativeGray)
-        {
-            return (isNonRelativeGray, isSelectedAndFocused: isSelected && isFocused) switch
+            => (isNonRelativeGray, isSelectedAndFocused: isSelected && isFocused) switch
             {
+                // Unselected revisions of the currently checked out branch and all revisions reachable from HEAD
                 (isNonRelativeGray: false, isSelectedAndFocused: false) => SystemColors.ControlText,
-                (isNonRelativeGray: false, isSelectedAndFocused: true) => SystemColors.HighlightText,
+
+                // Selected revisions of the currently checked out branch and all revisions reachable from HEAD
+                (isNonRelativeGray: false, isSelectedAndFocused: true) => ThemeModule.IsDarkTheme ? SystemColors.ControlText : SystemColors.HighlightText,
+
+                // All revisions not reachable from HEAD, revisions aren't selected
                 (isNonRelativeGray: true, isSelectedAndFocused: false) => SystemColors.GrayText,
 
                 // (isGray: true, isSelected: true)
                 _ => _highlightedGrayTextColor
             };
-        }
 
         private Color GetCommitBodyForeground(bool isSelected, bool isNonRelativeGray)
-        {
-            return (isNonRelativeGray, isSelected) switch
+            => (isNonRelativeGray, isSelected) switch
             {
+                // Unselected revisions of the currently checked out branch and all revisions reachable from HEAD
                 (isNonRelativeGray: false, isSelected: false) => SystemColors.GrayText,
+
+                // Selected revisions of the currently checked out branch and all revisions reachable from HEAD
                 (isNonRelativeGray: false, isSelected: true) => _highlightedGrayTextColor,
+
+                // All revisions not reachable from HEAD, revisions aren't selected
                 (isNonRelativeGray: true, isSelected: false) => _grayTextColor,
 
                 // (isGray: true, isSelected: true)
                 _ => _highlightedGrayTextColorCustom
             };
-        }
 
         private Brush GetBackground(bool isSelected, bool isFocused, int rowIndex, GitRevision? revision)
         {
@@ -411,11 +417,13 @@ public void Clear()
 
             // Reload settings that will be used during drawing
             _revisionGraphDrawNonRelativesTextGray = AppSettings.RevisionGraphDrawNonRelativesTextGray;
-            _highlightedGrayTextColor = getHighlightedGrayTextColor();
+
+            float degreeOfGrayness = ThemeModule.IsDarkTheme ? 6.0f : 1.4f;
+            _highlightedGrayTextColor = GetHighlightedGrayTextColor(degreeOfGrayness);
 
             // Note: The adaptive color fails here, step around 1.17f 145 -> 80
-            _grayTextColor = ThemeModule.IsDarkTheme ? Color.FromArgb(118, 118, 118) : getGrayTextColor(degreeOfGrayness: 1.4f);
-            _highlightedGrayTextColorCustom = getHighlightedGrayTextColor(degreeOfGrayness: 1.4f);
+            _grayTextColor = ThemeModule.IsDarkTheme ? Color.FromArgb(118, 118, 118) : GetGrayTextColor(degreeOfGrayness: 1.4f);
+            _highlightedGrayTextColorCustom = GetHighlightedGrayTextColor(degreeOfGrayness);
             _highlightAuthoredRevisions = AppSettings.HighlightAuthoredRevisions;
             _revisionGraphDrawAlternateBackColor = AppSettings.RevisionGraphDrawAlternateBackColor;
 
@@ -1000,14 +1008,14 @@ private void InitFonts()
             _monospaceFont = AppSettings.MonospaceFont;
         }
 
-        private static Color getHighlightedGrayTextColor(float degreeOfGrayness = 1f) =>
+        private static Color GetHighlightedGrayTextColor(float degreeOfGrayness = 1f) =>
             ColorHelper.GetHighlightGrayTextColor(
                 backgroundColorName: KnownColor.Control,
                 textColorName: KnownColor.ControlText,
                 highlightColorName: KnownColor.Highlight,
                 degreeOfGrayness);
 
-        private static Color getGrayTextColor(float degreeOfGrayness = 1f) =>
+        private static Color GetGrayTextColor(float degreeOfGrayness = 1f) =>
             ColorHelper.GetGrayTextColor(textColorName: KnownColor.ControlText, degreeOfGrayness);
     }
 }
```

Ideally, the revision grid colours should be made customisable (i.e., not being hardcoded to SystemColors), but I don't have enough time to implement that.

I also think that the dark+ theme should be folded into the dark theme.

![image](https://github.com/user-attachments/assets/53fcd404-0d6b-496c-a855-6309be61c72e)

RussKie commented:
I've been running this branch for some time, and I'm really impressed and quite happy with it. I think we should merge it in :)

gerhardol commented:
> @gerhardol I played tinkered with the dak theming, and I propose the following changes:

> Ideally, the revision grid colours should be made customisable (i.e., not being hardcoded to SystemColors), but I don't have enough time to implement that.
> 
> I also think that the dark+ theme should be folded into the dark theme.

Thanks, will take a look at the suggestions.
I want to keep this PR dark mode only, not changing the light mode at all, so I may separate the PR.

`dark` is mostly system colors and is a little gray, hard to get contrasts, `dark+` has slightly darker backgrounds similar to VS Code themes. dark has a few issues like tab headers so I prefer to keep it as a "bonus".

I also have the following question:
https://github.com/gitextensions/gitextensions/pull/12111#discussion_r1986163907
The terminology should maybe be cleaned up to use light/dark _mode_ rather than _themes_.

Similar, should there be a popup warning about dark mode is experimental in .net and GE?

RussKie commented:
> Similar, should there be a popup warning about dark mode is experimental in .net and GE?

We can have a warning in the release notes, I guess.

I vaguely remember bouncing ideas of introducing experimental settings, but I can't recall whether anything was implemented...

pmiossec commented:
> I've been running the dark+ as I found it more eye-pleasing, hence my suggestion to merge it into the dark theme.

I use it also since some days on daily basis (`dark` theme) and I found it good enough to be merged also.
There are very few places where there are problems.

But one example ("Active" is not readable):
![image](https://github.com/user-attachments/assets/8096e3d9-4f18-4414-9c9e-bfee2c03ddc5)

One "painful" problem is when you open an older version and you get this huge popup at startup 
![image](https://github.com/user-attachments/assets/fed7e949-8465-404e-ba39-91d2ce245e26)
Maybe not a problem if there is no move back. But I don't know if we should release a bugfix version handling this case gracefully...

gerhardol commented:
> > I've been running the dark+ as I found it more eye-pleasing, hence my suggestion to merge it into the dark theme.

I much prefer dark+ too, but got some negative feedback for similar changes for light mode, so I wanted to separate the setup.
The names matches VS Code, so I hope that is understandable.
But dark could be dark-system or so too, as a reference

> We can have a warning in the release notes, I guess.
I vaguely remember bouncing ideas of introducing experimental settings, but I can't recall whether anything was implemented...

Good, no popup then. Release notes is fine for me.
We had an experimental setting in Settings (no longer there), but here it is one of many themes.

> I use it also since some days on daily basis (`dark` theme) and I found it good enough to be merged also. There are very few places where there are problems.
> But one example ("Active" is not readable):

This is listed in the description - I have not found a workaround
There are more issues

> One "painful" problem is when you open an older version and you get this huge popup at startup

This is probably a problem mostly for developers, switching releases.
When this PR is merged the problem will no longer be seen when working with PRs (but still when testing behavior in older releases), so not seen so much.

pmiossec commented:
> There are more issues

I failed to remember one that I found earlier but it came back in my mind: we can't see anymore the selection when doing a blame.

Dark.css:
![image](https://github.com/user-attachments/assets/2fb57538-1324-4ee9-9266-47a792c5db0d)

in previous version:
![image](https://github.com/user-attachments/assets/2122890e-0847-4ec2-a550-605b203c955a)

mstv commented:
> I vaguely remember bouncing ideas of introducing experimental settings

https://github.com/gitextensions/gitextensions/blob/a10de0357204343d3b85dfdd6ef9b6a1080c5cf3/src/app/GitCommands/Settings/AppSettings.cs#L41

Currently not used because the regarding settings left the experimental state. Added in https://github.com/gitextensions/gitextensions/commit/12da3aa6a22ad27f0ace49927aa814bc3ce9ca79.

mstv commented:
(I have no feedback to this PR because my monitors and my eyes work much better in light mode.)

gerhardol commented:
Pushed a few fixes to NBug and suppress WFO5001
https://github.com/gitextensions/gitextensions/pull/12111#issuecomment-2709403043 is not included yet

RussKie commented on src/app/BugReporter/BugReportForm.cs:
```diff
@@ -77,6 +78,7 @@ public BugReportForm()
 
             // ToDo: Displaying report contents properly requires some more work.
             mainTabs.TabPages.Remove(mainTabs.TabPages["reportContentsTabPage"]);
+            Load += (s, e) => ((Form)s!).FixVisualStyle();
```
Is this the fix for this?

![image](https://github.com/user-attachments/assets/c5f36c9d-14fd-4cf3-b5ad-bdc33ddc9008)

gerhardol replied:
Yes
The themefix is basically the override of bugs in .net, where other general issues should be added.

(It was on my mental list but I had forgotten to add it in the description)

RussKie commented on src/app/BugReporter/BugReportForm.cs:
```diff
@@ -77,6 +78,7 @@ public BugReportForm()
 
             // ToDo: Displaying report contents properly requires some more work.
             mainTabs.TabPages.Remove(mainTabs.TabPages["reportContentsTabPage"]);
+            Load += (s, e) => ((Form)s!).FixVisualStyle();
```
```suggestion
            Load += (s, e) => this.FixVisualStyle();
```

Generally, it's better to override `OnLoad` as that avoid allocations for the event and the eventhandler.

gerhardol replied:
changed, also for the existing call

gerhardol commented:
> > I failed to remember one that I found earlier but it came back in my mind: we can't see anymore the selection when doing a blame.
> 
> I think it is still there, just the contrast is very low. I think we'll be tuning and fine-tuning the colours for quite some time after we merge this.

value is set now, adjusted to the background
Slightly more visible in dark+

gerhardol commented:
> @gerhardol I played tinkered with the dak theming, and I propose the following changes:

Adapted these suggestions some, the body text in the grid was not set correctly for instance
Also changed FileStatusList

For AuthoredHighlight you had changed the blueish color to gray. I find that still to close to selection colors and set a slightly yellow color. There is not much contrast to play with.

Plan to update screenshots tomorrow
I believe this is ready to merge, to start the finetuning

gerhardol commented:
In light mode, the subject/body for the commits in n branches not merged to master are darker, basically the same in dark mode. Will change that first. There is less contrast to play with, will be less visible though.

gerhardol commented:
Tweaked the colors some more, I believe this is good enough for usage (and more tweaks).
Will try to compile a list of issues to the Winforms team when this is merged too.

> I've been running the dark+ as I found it more eye-pleasing, hence my suggestion to merge it into the dark theme.

I want to keep dark theme as a "system compliant" and dark+ as some stretches to the design that is maybe not as consistent, compare to VS Code themes.
For instance tab headers are not adjusted in dark+ (not a big problem), some popups differ to the background. But dark+ is what I use myself.

RussKie commented:
@gerhardol, FWIW this could be useful https://devblogs.microsoft.com/dotnet/introducing-winforms-analyzers/.

gerhardol commented:
A short introduction and issues listed in the wiki
https://github.com/gitextensions/gitextensions/wiki/Dark-Theme

In addition, I renamed files so the repo can be cloned in Windows and applied markdownlint and did a minor cleanup

RussKie commented:
@KlausLoeffelmann has shared some news with me regarding the dark theming in .NET 10. 

The API and implementations are expected to be more stable, though those will never reach a perfect state, unfortunately.
The `TabControl` is expected to work nicely in the default mode (i.e., tabs at the top); the `ListView` headers should now render correctly, although the `GridLines` could be too bright; the scrollbars will look native; the `Buttons` will look correctly in every scaling (96 dpi did not work before); and the `ToolStrip` will have a DarkMode renderer.

The Windows Forms will be looking for more feedback from us, when we start moving to the .NET 10 bits.

gerhardol commented:
> The Windows Forms will be looking for more feedback from us, when we start moving to the .NET 10 bits.

@KlausLoeffelmann (@russkie may have to forward)
My feedback and info to users is here: https://github.com/gitextensions/gitextensions/wiki/Dark-Mode

I guess most where we have a workaround are fixed in .net10
(they are not critical for us though, like the buttons work fine...)
https://github.com/gitextensions/gitextensions/wiki/Dark-Mode#workarounds

The urgent are like the list categories:
https://github.com/gitextensions/gitextensions/wiki/Dark-Mode#other

The application colors could be improved too, not related to .net.

RussKie commented:
Here's a short animation of the app on .NET 10 preview7 (kudos to @KlausLoeffelmann for making it):

https://gist.github.com/user-attachments/assets/2e5b0038-a252-41eb-8bbd-fd04f6fa08e0

gerhardol commented:
> Here's a short animation of the app on .NET 10 preview7 (kudos to @KlausLoeffelmann for making it):

* I do not see any categories in FileLists, that is where there is no workaround now.
* (Are tooltips darker?)
* (Disabled button text visible now?)
* Some read-only boxes are still bright
* Obviously no checks for issues where GE has a workaround

@KlausLoeffelmann , some patches coming too, hinted in the screenshot?

App related:

* The most visual issue is the ConEmu theme. We could maybe make terminal default (then a popup is seen in some situations though) or implement to default follow the GE theme.
* Why are settings summary gray? (colors are a little too bright now, a compromise with light theme)

gerhardol commented:
@KlausLoeffelmann  Thanks for this

>> I do not see any categories in FileLists, that is where there is no workaround now.
> That's the most challenging one. Aiming for RC1, but cannot promise.

If we have something that can be overridden (maybe there is a way already?), I would be happy.

> > * (Disabled button text visible now?)
> > * Some read-only boxes are still bright
> 
> * Fix is underway. Should only be the ComboBox controls, which need almost the same approach as the TextBoxes.

Yes probably only ComboBoxes

> > * Obviously no checks for issues where GE has a workaround
> 
> Not sure what you mean: Can you elaborate or screenshot?

GE overrides some controls. The wiki points to this source file:
https://github.com/gitextensions/gitextensions/blob/master/src/app/GitExtUtils/GitUI/Theming/ThemeFix.cs

So buttons are not really a problem as they are set to Flat, ToolStripLabel colors overridden etc.

> > @KlausLoeffelmann , some patches coming too, hinted in the screenshot?
> 
> This is the current nightly-public-build Preview 7, but I cannot tell you exactly, which already merged fixes are still out. They should surface the coming days.

The video hinted that you had made fixes too.

> > App related:
> 
> > * The most visual issue is the ConEmu theme. We could maybe make terminal default (then a popup is seen in some situations though) or implement to default follow the GE theme.
> 
> Is that a comment in my direction? If it is, please give me a bit more info to understand what you mean! Thanks!

No, there are of course changes needed to be tweaked in the colors GE uses.
In this case ConEmu should have a better default theme, or the WinForms standard controls be used.

> > * Why are settings summary gray? (colors are a little too bright now, a compromise with light theme)
> 
> UpDown Control was a surprise, btw. It should not have rendered with the old button style, AND on top not in dark mode. That one is a regression, and I need to find out, what's going on there.

This is not updown, those are normal buttons that are styled.
(colors should be tweaked in dark mode)

![image](https://github.com/user-attachments/assets/77c398a5-f301-40a3-993b-5710582a5b98)

colors are set directly:

````cs
       /// <summary>
        /// Renders settings as correctly configured.
        /// </summary>
        private static void RenderSettingSet(Button settingButton, Button settingFixButton, string text)
        {
            settingButton.BackColor = OtherColors.BrightGreen;
            settingButton.ForeColor = ColorHelper.GetForeColorForBackColor(settingButton.BackColor);
            settingButton.Text = text;
            settingFixButton.Visible = false;
        }
```

> 
> On top: Please note, that if you want a custom control (most of all if it has scroll bars which need to be themed), you need to override CreateParams, and then do

GE has very few such controls (CreateParams is used once now).
Most changes are in the logic, not the looks.
The filelists could be customized to conform better with other Win controls (but it is not so much off, IMO).

gerhardol commented:
@KlausLoeffelmann 
I took .net10.0 rc1 for a spinn, not much migration needed (if you saw other issues and care to push changes, please do...)
I created a draft PR, but there are months until it can be merged.
https://github.com/gitextensions/gitextensions/pull/12515
